File: offer-search-service/src/main/java/pl/jugademy/offersearch/infrastructure/KafkaOfferSubscriber.kt
package pl.jugademy.offersearch.infrastructure

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.slf4j.LoggerFactory
import pl.jugademy.offersearch.domain.OfferEvent
import pl.jugademy.offersearch.domain.OfferSubscriber
import java.util.concurrent.ExecutorService

class KafkaOfferSubscriber(
        private val consumer: KafkaConsumer<ByteArray, ByteArray>,
        private val topic: String,
        private val mapper: ObjectMapper,
        private val executor: ExecutorService
) : OfferSubscriber {

    override fun subscribe(consume: (offer: OfferEvent) -> Unit) {
        executor.execute {
            consumer.subscribe(setOf(topic))
            while (true) {
                val records = consumer.poll(100)
                records.forEach { event ->
                    logger.info("Received event with offset ${event.offset()}")
                    val offerEvent = mapper.readValue(event.value(), OfferEvent::class.java)
                    consume(offerEvent)
                }
            }
        }
    }

    companion object {
        val logger = LoggerFactory.getLogger(KafkaOfferSubscriber::class.java)!!
    }
}

File: offer-cache-service/src/main/scala/pl/jugademy/offercache/infrastructure/KafkaOfferSubscriber.scala
package pl.jugademy.offercache.infrastructure

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.slf4j.LoggerFactory
import pl.jugademy.offercache.domain.{OfferEvent, OfferSubscriber}

import scala.concurrent.ExecutionContext
import collection.JavaConverters._

class KafkaOfferSubscriber(
  consumer: KafkaConsumer[Array[Byte], Array[Byte]],
  topic: String,
  mapper: ObjectMapper
)(implicit ec: ExecutionContext) extends OfferSubscriber {

  private val logger = LoggerFactory.getLogger(classOf[KafkaOfferSubscriber].getName)

  override def subscribe(consume: OfferEvent => Unit): Unit =
    ec.execute {
      () => {
        consumer.subscribe(Set(topic).asJava)
        while (true) {
          val records = consumer.poll(100)
          records.forEach { rec =>
            logger.info(s"Received event with offset ${rec.offset()}")
            try {
              val offerEvent = mapper.readValue(rec.value(), classOf[OfferEvent])
              consume(offerEvent)
            } catch {
              case e: Exception =>
                logger.error(s"Could not deserialize event with offset ${rec.offset()} due to ${e.getMessage}", e)
            }
          }
        }
      }
    }
}
